ext/GPU/CUDAKernelLauncher.jl

```diff
@@ -0,0 +1,139 @@
+## This file manifsts the launch of GPU kernel on CUDA backend ##
+
+"""
+    Ferrite.init_kernel(::Type{BackendCUDA}, n_cells::Ti, n_basefuncs::Ti, kernel::Function, args::Tuple) where {Ti <: Integer}
+
+Initialize a CUDA kernel for the Ferrite framework.
+
+# Arguments
+- `::Type{BackendCUDA}`: Specifies the CUDA backend.
+- `n_cells::Ti`: Number of cells in the problem.
+- `n_basefuncs::Ti`: Number of shape functions per cell.
+- `kernel::Function`: The CUDA kernel function to execute.
+- `args::Tuple`: Tuple of arguments for the kernel.
+
+# Returns
+- A `LazyKernel` object encapsulating the kernel and its execution configuration.
+
+# Errors
+Throws an `ArgumentError` if CUDA is not functional (e.g., due to missing drivers or improper installation).
+"""
+function Ferrite.init_kernel(::Type{BackendCUDA}, n_cells::Ti, n_basefuncs::Ti, kernel::Function, args::Tuple) where {Ti <: Integer}
+    if CUDA.functional()
+        return LazyKernel(n_cells, n_basefuncs, kernel, args, BackendCUDA)
+    else
+        throw(ArgumentError("CUDA is not functional, please check your GPU driver and CUDA installation"))
+    end
+end
+
+"""
+    Ferrite.launch!(kernel::LazyKernel{Ti, BackendCUDA}) where {Ti}
+
+Launch a CUDA kernel encapsulated in a `LazyKernel` object.
+
+# Arguments
+- `kernel::LazyKernel`: The kernel to be launched, along with its configuration.
+
+# Returns
+- `nothing`: Indicates that the kernel was launched and synchronized successfully.
+"""
+function Ferrite.launch!(kernel::LazyKernel{Ti, BackendCUDA}) where {Ti}
+    n_cells = kernel.n_cells
+    n_basefuncs = kernel.n_basefuncs
+    ker = kernel.kernel
+    args = kernel.args
+    kernel = @cuda launch = false ker(args...)
+    config = launch_configuration(kernel.fun)
+    threads = convert(Ti, min(n_cells, config.threads, 256))
+    shared_mem = _calculate_shared_memory(threads, n_basefuncs)
+    blocks = _calculate_nblocks(threads, n_cells)
+
+    ## use dynamic shared memory if possible
+    _can_use_dynshmem(shared_mem) && return CUDA.@sync kernel(args...; threads, blocks, shmem = shared_mem)
+
+    ## otherwise use global memory
+    nes = blocks * threads
+    kes = CUDA.zeros(Float32, nes, n_basefuncs, n_basefuncs)
+    fes = CUDA.zeros(Float32, nes, n_basefuncs)
```

Comment:
Okay, this actually a tough design issue that I have faced and to be honest I greatly improvised on this one, but essentially the local matrices are being transferred in two steps:
1. encapsulate them in a new dohandler that acts as a wrapper only and replace the standard dofhandler that is passed in the arguments by the new one -> `_to_localdh`

```julia
  kes = CUDA.zeros(Float32, nes, n_basefuncs, n_basefuncs)
    fes = CUDA.zeros(Float32, nes, n_basefuncs)
    args = _to_localdh(args, kes, fes) ## add the kes and fes to the new dof handler
    CUDA.@sync @cuda blocks = blocks threads = threads ker(args...)
```
2. 
```julia
struct LocalsGPUDofHandler{DH <: AbstractDofHandler, LOCAL_MATRICES, LOCAL_VECTORS} <: AbstractGPUDofHandler
    dh::DH
    Kes::LOCAL_MATRICES
    fes::LOCAL_VECTORS
end
```
2. since the dohandler is already replaced before executing the kernel, the celliterator that is going to be invoked is the following one, which basically access the right partition of the local stiffness matrix and load vector and pass it to the cell cache.
```julia
function Ferrite.CellIterator(dh_::Ferrite.LocalsGPUDofHandler, ::Int32)
    dh = dh_ |> dofhandler
    grid = get_grid(dh)
    n_cells = grid |> getncells |> Int32
    bd = blockDim().x
    local_thread_id = threadIdx().x
    global_thread_id = (blockIdx().x - Int32(1)) * bd + local_thread_id
    ke = cellke(dh_, global_thread_id)
    fe = cellfe(dh_, global_thread_id)
    return CUDAGlobalCellIterator(dh, grid, n_cells, ke, fe, local_thread_id)
end
```